### DISCUSSION TITLE
Where are cookies and localstorage data stored

### DISCUSSION THREAD

**Discussion by pokecheater** [tauri-apps/tauri#8637] [Q&A]
Hey Tauri-Team :),

Since this feature https://github.com/tauri-apps/wry/issues/518 has not been implemented for almost 2 years since the first request, I need a workaround for this. If I can access the cookie storage by myself inside the rust backend everything would be fine, but, inside the docs, I can't find out where they are stored.

I tried to find it out with the following code snippet:

```javascript
import { appDataDir, dataDir, localDataDir } from '@tauri-apps/api/path';
const appDataDirPath = await appDataDir();
console.log("appDataDirPath", appDataDirPath);
```

and  it states:
[Log] appDataDirPath – "/Users/c3po/Library/Application Support/XXX-app/"

but that path does not exist.

I tried to set a custom data_directory as proposed in https://github.com/tauri-apps/wry/issues/518 but that does not work either since the path gets created but remains empty all the time:

```javascript
tauri::Builder::default()
        .setup(|app| {
            let path_str = "/Users/c3po/session_data/cookies";
            let data_directory: PathBuf = PathBuf::from(path_str);
            let path_string: String = data_directory.to_string_lossy().to_string();
            let window = tauri::WindowBuilder::new(app, "label", tauri::WindowUrl::App("index.html".into()))
            .data_directory(data_directory)
            .build()?;

            // let cookies = std::fs::read_to_string(path_string).expect("could not read cookies");
            // println!("{cookies}");
            Ok(())
        }) 
```

Can anyone help me here out?
Thx in advance <3.

**Comment by FabianLars** [comment]
`/Users/c3po` means you're on macos right? We can't control the data dir on macos so that's probably why it's empty, though i would have expected that i won't be created at all.
macos/wkwebview will store its data in /Users/c3po/Library/WebKit/<bundle-id>/` (or appname instead of bundleid in `tauri dev` mode). Though i'm not sure what format(s) it uses. On Linux for example cookies are stored as plain text, on Windows it's an sqlite database.
There's also `/Users/c3po/Library/Cookies/` but i think the webview ones are not part of that, not sure 🤔

**Comment by xyanyue** [comment]
I have a code that supports Windows and Mac implementation through `PlatformWebview`. If you want to implement it yourself, you can refer to it
https://zhuanlan.zhihu.com/p/13429407786

**Reply by codewithbobb** [reply]
I've been struggling for a while, but the code still won't run.
Would you kindly send me a simple sample project so I can study it?
Thánks so much

**Reply by xyanyue** [reply]
> I've been struggling for a while, but the code still won't run. Would you kindly send me a simple sample project so I can study it? Thánks so much

```
pub async fn set_window_cookie(
    webview: &tauri::WebviewWindow,
    cookies: Vec<Cookie>,
) -> Result<(), AppError<&'static str>> {
    // for cookie in &cookies {
    //     let javascript = format!(
    //         "document.cookie = '{}={}; path={};';",
    //         cookie.name, cookie.value, cookie.path
    //     );
    //     let _ = webview.eval(javascript.as_str())?;
    // }
    webview.with_webview(move |webview| {
        #[cfg(windows)]
        unsafe {
            let core = webview.controller().CoreWebView2().unwrap();
            let core2 = Interface::cast::<ICoreWebView2_2>(&core).unwrap();
            // let uri = HSTRING::from("");
            let manager = core2.CookieManager().unwrap();
            for cookie in cookies {
                if let Ok(cookie_w) = manager.CreateCookie(
                    pwstr_from_str(&cookie.name),
                    pwstr_from_str(&cookie.value),
                    pwstr_from_str(&cookie.domain),
                    pwstr_from_str(&cookie.path),
                ) {
                    if let Err(e) = manager.AddOrUpdateCookie(&cookie_w) {
                        error!("cookie 写入失败:{}", e)
                    }
                }
            }
        }
        #[cfg(target_os = "macos")]
        unsafe {
            let b =
                move |result: cocoa::base::id, _: cocoa::base::id| println!("JS设置COOKIE 成功");
            let b = block::ConcreteBlock::new(b);
            let b = b.clone();
            let web_obj = webview.inner();
            let cookie_store = web_obj.configuration().websiteDataStore().httpCookieStore();
            for cookie in cookies {
                // println!("设置cookie:{:?}", &cookie);
                let cookie_ns = cookie_nsdict(
                    cookie.domain,
                    cookie.path.clone(),
                    cookie.name.clone(),
                    cookie.value.clone(),
                );

                cookie_store.setCookie(create_cookie(cookie_ns));
            }
        }
    })?;
    Ok(())
}
```
原理就是通过`PlatformWebview ` 去调用各个平台的webview的API

MacOS 可以参考`https://crates.io/crates/darwin-webkit`，然后通过Apple webview 开发文档，添加cookie操作

注意：使用的 create `windows` 、`webview2-com` 、`objc` 、`cocoa` 和 tauri 中使用的版本对应，可以去tauri的源码中的Cargo.toml 查看。

新版tauri好像已经提供了，读取cookie的操作
